### ISSUE TITLE
Evicting the cache for a query with both @include and @nonreactive directives can cause errors if variables have been changed

### ISSUE THREAD

**Issue by jbachhardie** [apollographql/apollo-client#11987]
### Issue Description

This one took quite a while to get a minimal reproduction for since a lot of things have to be true for it to happen so I'll try to explain it all step by step as well as attaching a Codesandbox. If you:

1. have a query that uses `@include` with a variable that is set through a query-level default
2. and which also uses `@nonreactive` somewhere in it
3. and you're calling it with `useQuery`
4. and you change the variables to something that doesn't include the variable that is used in the `@include`
5. and you then evict a field of the query using `cache.evict`
6. and you then call `cache.gc`
7. then Apollo errors with a "Invalid variable referenced in @include directive" error, even though the variable has a valid default

This bubbles up as an unhandled error that can't be caught by either `try/catch` around the `cache.gc()` call or React error boundaries, which made it completely crash out our React Native app.

I know a little bit of what is happening here, [observable.reobserve() in useResubscribeIfNecessary](https://github.com/apollographql/apollo-client/blob/c95848e859fb7ce0b3b9439ac71dff880f991450/src/react/hooks/useQuery.ts#L540) sets new variables on the observable but doesn't apply query defaults to them because that logic is in the [QueryManager](https://github.com/apollographql/apollo-client/blob/c95848e859fb7ce0b3b9439ac71dff880f991450/src/core/QueryManager.ts#L713) rather than the observable. This normally doesn't cause any problems AFAICT but if the [query has a nonreactive directive](https://github.com/apollographql/apollo-client/blob/c95848e859fb7ce0b3b9439ac71dff880f991450/src/core/ObservableQuery.ts#L331) then the Observable has to process the selection sets with the new variables, which are missing their defaults. I'm not super sure why this only happens if `cache.gc` is called, though, and I don't immediately know how I'd refactor things to fix this in a PR so I'm just reporting it for now.

### Link to Reproduction

https://codesandbox.io/p/devbox/upbeat-bush-mwmqx6?workspaceId=03104bf9-16b8-430d-a030-08b693449298

### Reproduction Steps

Click "Change variables" and then "Evict cache" in the CodeSandbox, see the uncaught error in the console

### `@apollo/client` version

3.10.5

**Comment by phryneas** [comment]
Hi Jae,
thank you for that thorough analysis, this is great!

I believe there is no beautiful fix for this - we'll just have to add variable defaults everywhere they're missing.
I've started a PR over in #11990 - if you have the time, please try out the PR build and report back if it fixes the issue for you :)

**Comment by phryneas** [comment]
This should be solved by #12559, which we will release as part of Apollo Client 4.0. Should be out in the next alpha soon.

**Comment by jerelmiller** [comment]
Hey @jbachhardie 👋 

I can confirm that [4.0.0-alpha.10](https://github.com/apollographql/apollo-client/releases/tag/v4.0.0-alpha.10) fixes the issue. Please see my updated codesandbox of your reproduction to see this working: https://codesandbox.io/p/devbox/hungry-http-n5gzqk?file=%2Fsrc%2Flink.js%3A28%2C22.

Since that fix is merged and published, I'm going to go ahead and close this issue. We'd appreciate you trying out the alpha if you get a chance, but otherwise this will be released when we get 4.0 out the door. Thanks!

**Comment by jbachhardie** [comment]
> Hey @jbachhardie 👋 
> 
> I can confirm that [4.0.0-alpha.10](https://github.com/apollographql/apollo-client/releases/tag/v4.0.0-alpha.10) fixes the issue. Please see my updated codesandbox of your reproduction to see this working: https://codesandbox.io/p/devbox/hungry-http-n5gzqk?file=%2Fsrc%2Flink.js%3A28%2C22.
> 
> Since that fix is merged and published, I'm going to go ahead and close this issue. We'd appreciate you trying out the alpha if you get a chance, but otherwise this will be released when we get 4.0 out the door. Thanks!

Thank you. It's a bit tricky since we don't have the case that caused this in the codebase anymore. I did ask someone to look into reproducing it and see if the branch solved it but it must've got reprioritized. I'll see if I can look into it but I'm guessing if it works in the sandbox it'll work in production, it was that exact error.

**Comment by jerelmiller** [comment]
Yep I'm pretty confident it will 🙂. Any variable value with an `undefined` value is now re-evaluated against the query so any default vars will be filled in. If your reproduction was an accurate representation of the problem, then I feel confident its fixed. Thanks for reporting this!
